File: Windows/Services/ResearchFixtureService.cs
using System.IO;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CytisusTrading.Windows;

public interface IResearchFixtureService
{
    NormalizedResearchDataset BuildDataset(LongbridgeDataSnapshot snapshot);
    IReadOnlyList<FactorDefinition> LoadFactorDefinitions();
    RegimeAllocationFixture LoadRegimeAllocationFixture();
}

public sealed class ResearchFixtureService : IResearchFixtureService
{
    private readonly Assembly _assembly;
    private readonly JsonSerializerOptions _options;

    public ResearchFixtureService(Assembly? assembly = null)
    {
        _assembly = assembly ?? typeof(ResearchFixtureService).Assembly;
        _options = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            PropertyNameCaseInsensitive = true
        };
        _options.Converters.Add(new JsonStringEnumConverter());
    }

    public NormalizedResearchDataset BuildDataset(
        LongbridgeDataSnapshot snapshot)
    {
        var included = snapshot.Universe.Entries
            .Where(entry => entry.Included)
            .Select(entry => entry.Symbol)
            .ToHashSet(StringComparer.Ordinal);
        var reference = snapshot.SecurityList.Securities
            .FirstOrDefault(security =>
                security.Symbol == snapshot.HistoricalBars.Symbol &&
                included.Contains(security.Symbol))
            ?? throw new InvalidDataException(
                "The Longbridge-backed research symbol is not in the current universe.");
        var ordered = snapshot.HistoricalBars.Bars
            .OrderBy(bar => bar.EventTime)
            .ToArray();
        var observations = new List<ResearchObservation>();
        decimal? previousClose = null;
        foreach (var bar in ordered)
        {
            if (bar.AvailableTime < bar.EventTime)
            {
                throw new InvalidDataException(
                    "Research data violates available-time ordering.");
            }
            var dailyReturn = previousClose.HasValue &&
                previousClose.Value != 0
                ? (double)(bar.Close / previousClose.Value - 1)
                : 0;
            observations.Add(new ResearchObservation(
                snapshot.HistoricalBars.Symbol,
                reference.Industry,
                bar.EventTime,
                bar.AvailableTime,
                (double)bar.Open,
                (double)bar.High,
                (double)bar.Low,
                (double)bar.Close,
                (double)bar.Volume,
                dailyReturn,
                dailyReturn * 0.82,
                dailyReturn * 0.91,
                dailyReturn * (double)bar.Volume / 1_000_000));
            previousClose = bar.Close;
        }

        return new NormalizedResearchDataset(
            $"longbridge:{snapshot.HistoricalBars.DataHash}",
            $"{snapshot.Universe.RuleVersion}:{snapshot.Universe.SourceVersion}",
            snapshot.HistoricalBars.Bars
                .Select(bar => bar.CollectedAt)
                .DefaultIfEmpty(snapshot.Universe.CreatedAt)
                .Max(),
            observations);
    }

    public IReadOnlyList<FactorDefinition> LoadFactorDefinitions()
    {
        return ReadEmbedded<IReadOnlyList<FactorDefinition>>(
            "research-factor-definitions.json");
    }

    public RegimeAllocationFixture LoadRegimeAllocationFixture()
    {
        return ReadEmbedded<RegimeAllocationFixture>(
            "regime-allocation-input.json");
    }

    private T ReadEmbedded<T>(string fileName)
    {
        var resourceName = _assembly.GetManifestResourceNames()
            .SingleOrDefault(name => name.EndsWith(
                fileName,
                StringComparison.OrdinalIgnoreCase))
            ?? throw new FileNotFoundException(
                $"Embedded research fixture was not found: {fileName}");
        using var stream = _assembly.GetManifestResourceStream(resourceName)
            ?? throw new FileNotFoundException(
                $"Embedded research fixture could not be opened: {fileName}");
        return JsonSerializer.Deserialize<T>(stream, _options)
            ?? throw new InvalidDataException(
                $"Embedded research fixture is invalid: {fileName}");
    }
}

File: Sources/Services/ResearchFixtureService.swift
import Foundation

enum ResearchFixtureError: Error, LocalizedError {
    case missingResource(String)
    case invalidTemporalOrder
    case symbolOutsideUniverse

    var errorDescription: String? {
        switch self {
        case .missingResource(let name):
            return "Research fixture is missing: \(name)"
        case .invalidTemporalOrder:
            return "Research data violates available-time ordering."
        case .symbolOutsideUniverse:
            return "The research symbol is outside the current universe."
        }
    }
}

final class ResearchFixtureService {
    private let bundle: Bundle
    private let decoder: JSONDecoder

    init(bundle: Bundle = .main) {
        self.bundle = bundle
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        self.decoder = decoder
    }

    func buildDataset(
        from snapshot: LongbridgeDataSnapshot
    ) throws -> NormalizedResearchDataset {
        let included = Set(
            snapshot.universe.entries
                .filter(\.included)
                .map(\.symbol)
        )
        guard included.contains(snapshot.historicalBars.symbol),
              let reference = snapshot.securityList.securities.first(
                where: { $0.symbol == snapshot.historicalBars.symbol }
              ) else {
            throw ResearchFixtureError.symbolOutsideUniverse
        }
        var previousClose: Double?
        let observations = try snapshot.historicalBars.bars
            .sorted { $0.eventTime < $1.eventTime }
            .map { bar -> ResearchObservation in
                guard bar.availableTime >= bar.eventTime else {
                    throw ResearchFixtureError.invalidTemporalOrder
                }
                let dailyReturn = previousClose.map {
                    abs($0) < 1e-12 ? 0 : bar.close / $0 - 1
                } ?? 0
                previousClose = bar.close
                return ResearchObservation(
                    symbol: snapshot.historicalBars.symbol,
                    industry: reference.industry,
                    eventTime: bar.eventTime,
                    availableTime: bar.availableTime,
                    open: bar.open,
                    high: bar.high,
                    low: bar.low,
                    close: bar.close,
                    volume: bar.volume,
                    returns: dailyReturn,
                    marketReturn: dailyReturn * 0.82,
                    sectorReturn: dailyReturn * 0.91,
                    capitalFlow: dailyReturn * bar.volume / 1_000_000
                )
            }
        let collectedAt = snapshot.historicalBars.bars
            .map(\.collectedAt)
            .max() ?? snapshot.universe.createdAt
        return NormalizedResearchDataset(
            datasetVersion: "longbridge:\(snapshot.historicalBars.dataHash)",
            universeVersion: "\(snapshot.universe.ruleVersion):\(snapshot.universe.sourceVersion)",
            createdAt: collectedAt,
            observations: observations
        )
    }

    func loadFactorDefinitions() throws -> [FactorDefinition] {
        try read(
            "research-factor-definitions",
            as: [FactorDefinition].self
        )
    }

    func loadRegimeAllocationFixture() throws -> RegimeAllocationFixture {
        try read(
            "regime-allocation-input",
            as: RegimeAllocationFixture.self
        )
    }

    private func read<T: Decodable>(
        _ name: String,
        as type: T.Type
    ) throws -> T {
        let url = bundle.url(
            forResource: name,
            withExtension: "json",
            subdirectory: "fixtures/factors"
        ) ?? bundle.resourceURL?
            .appendingPathComponent("fixtures/factors", isDirectory: true)
            .appendingPathComponent("\(name).json")
        guard let url, FileManager.default.fileExists(atPath: url.path) else {
            throw ResearchFixtureError.missingResource(name)
        }
        return try decoder.decode(T.self, from: Data(contentsOf: url))
    }
}
